### PR TITLE
VEIOSS-366 Only Allow Traders to Sequentially Submit Returns Based on  their Commencement Date

### PR DESCRIPTION
**Pull request by pbyrne84**
Links to https://github.com/hmrc/ioss-returns/pull/47

### CODE REVIEW

File: app/viewmodels/yourAccount/ReturnsViewModel.scala [hunk 1]
```diff
@@ -17,34 +17,74 @@
...
 case class ReturnsViewModel(
-                           contents: Seq[Paragraph],
-                           linkToStart: Option[LinkModel] = None
+                             contents: Seq[Paragraph],
+                             linkToStart: Option[LinkModel] = None
                            )
 
 object ReturnsViewModel {
 
   def apply(returns: Seq[Return])(implicit messages: Messages): ReturnsViewModel = {
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/ReturnsViewModel.scala:33]: Moved this logic to NextReturnCalculation so the filter and this model have the same basis for the next return.

File: app/viewmodels/yourAccount/NextReturnCalculation.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  val maybeExpectedCurrentReturn: Option[Return] =
+    overDueReturns
+      .sortBy(_.period.toEtmpPeriodString)
+      .headOption
+      .map(Some.apply)
+      .getOrElse(maybePendingReturn)
+
+}
+
+object NextReturnCalculation {
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/NextReturnCalculation.scala:39]: Logic from ReturnsViewModel which is used by filter as well

File: app/viewmodels/yourAccount/ReturnsViewModel.scala [hunk 3]
```diff
@@ -17,34 +17,80 @@
...
+              contents = Seq(nextReturnNoLinkParagraph(pendingReturn.period))
+            )
+
+          case otherReturn: OtherReturn =>
+            createModelFromDueReturn(waypoints, otherReturn)
+        }
+    }
+  }
+
+  private def nextReturnNoLinkParagraph(nextReturn: StandardPeriod)(implicit messages: Messages) =
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/ReturnsViewModel.scala:51]: Moved from bottom and NoLink added to name as this should not be a valid return for doing in the filter.

File: app/viewmodels/yourAccount/ReturnsViewModel.scala [hunk 4]
```diff
@@ -17,34 +17,80 @@
...
+            linkToStart = Some(startDueReturnLink(waypoints, dueReturn.period))
+          )
+        }.getOrElse {
+          ReturnsViewModel(
+            contents = Seq.empty,
+            linkToStart = None
+          )
+        }
+
+      case ::(onlyOverDueReturn, Nil) =>
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/ReturnsViewModel.scala:74]: Replaces 

```scala
   case (1, None, _) =>
        val contents = dueReturn.map(dueReturn =>
          Seq(returnOverdueSingularParagraph(), returnDueParagraph(dueReturn.period))).getOrElse(Seq(returnOverdueParagraph()))
        ReturnsViewModel(
          contents = contents,
          linkToStart = Some(startOverdueReturnLink(waypoints, overdueReturns.head.period))
        )
```

File: app/viewmodels/yourAccount/ReturnsViewModel.scala [hunk 5]
```diff
@@ -17,34 +17,80 @@
...
+        }.getOrElse(Seq(returnOverdueParagraph()))
+
+        ReturnsViewModel(
+          contents = contents,
+          linkToStart = Some(startOverdueReturnLink(waypoints, onlyOverDueReturn.period))
+        )
 
-    nextReturn.map(
-      nextReturn =>
+      case manyOverDueReturns =>
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/ReturnsViewModel.scala:84]: Replaces 

```scala

      case (x, None, _) =>
        val contents = dueReturn.map(dueReturn =>
            Seq(returnsOverdueParagraph(x), returnDueParagraph(dueReturn.period)))
          .getOrElse(Seq(onlyReturnsOverdueParagraph(x)))
        ReturnsViewModel(
          contents = contents,
          linkToStart = Some(startOverdueReturnLink(waypoints, overdueReturns.minBy(_.period.lastDay.toEpochDay).period))
        )
```

File: app/viewmodels/yourAccount/NextReturnCalculation.scala [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+sealed trait ActionableReturn
+
+case class NextReturn(pendingReturn: Return) extends ActionableReturn
+
+case class OtherReturn(maybeDueReturn: Option[Return], overDueReturns: Seq[Return])
+  extends ActionableReturn {
+
+  val overdueCount: Int = overDueReturns.size
+
+  val maybeExpectedCurrentReturn: Option[Return] =
```
Comment (pbyrne84) [inline app/viewmodels/yourAccount/NextReturnCalculation.scala:30]: Logic taken from ReturnsViewModel

We prioritise overdue and defer due

File: test/viewmodels/yourAccount/NextReturnCalculationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+import base.SpecBase
+import models.SubmissionStatus.{Complete, Due, Excluded, Next, Overdue}
+import models.{StandardPeriod, SubmissionStatus}
+import org.scalatest.EitherValues
+import org.scalatest.prop.TableDrivenPropertyChecks
+
+import java.time.YearMonth
+
+class NextReturnCalculationSpec extends SpecBase with EitherValues with TableDrivenPropertyChecks{
```
Comment (pbyrne84) [inline test/viewmodels/yourAccount/NextReturnCalculationSpec.scala:27]: Prioritisation logic based on existing ReturnViewModel

File: README.md [hunk 8]
```diff
@@ -68,6 +68,26 @@ An asterisk can be used as a wildcard character without having to enter the pack
...
+### To run the application locally execute the following:
+
+The service needs to run in testOnly mode in order to access the testOnly get-passcodes endpoint which will generate a passcode for email verification.
+```
+sm2 --stop IOSS_RETURNS_FRONTEND
+```
+and
+```
+sbt run -Dapplication.router=testOnlyDoNotUseInAppConf.Routes
+```
```
Comment (hmrcvmu) [inline README.md:88]: Look like the same section is in the README above

Comment (pbyrne84) [reply README.md:88]: We both seemed to have added it :)

File: app/controllers/actions/CheckIsCurrentReturnPeriodFilter.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                                          (implicit protected val executionContext: ExecutionContext)
+  extends ActionFilter[OptionalDataRequest] {
+
+  override protected def filter[A](request: OptionalDataRequest[A]): Future[Option[Result]] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+
+    val iossNumber = request.iossNumber
+    returnStatusConnector.getCurrentReturns(iossNumber).map { errorOrReturns: CurrentReturnsResponse =>
+      errorOrReturns match {
+        case Left(value: ErrorResponse) => throw new RuntimeException(s"${value}")
```
Comment (hmrcvmu) [inline app/controllers/actions/CheckIsCurrentReturnPeriodFilter.scala:45]: could use `value.toString` or `s"$value"`

**Review by hmrcvmu — APPROVED** [review]

**Pull request merged**